torch/csrc/jit/passes/common_subexpression_elimination.cpp

```diff
@@ -11,87 +11,110 @@ namespace torch {
 namespace jit {
 namespace {
 
-// The function implements common subexpression elimination.
-// Since the nodes are visited in topological order, one pass is enough.
-void EliminateCommonSubexpression(
-    Block* block,
-    const AliasDb& aliasDb,
-    std::function<Node*(Node*)> parent_lookup_fn) {
-  std::unordered_set<Node*, HashNode, EqualNode> subexprs;
-  for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
-    auto node = *it;
-    if (node->hasSideEffects()) {
-      GRAPH_DEBUG("Node was skipped due to side effects:\n", *node);
-      continue;
-    }
-    if (node->isNondeterministic()) {
-      GRAPH_DEBUG("Node was skipped due to its non determinism:\n", *node);
-      continue;
-    }
-    if (aliasDb.hasWriters(node)) {
-      GRAPH_DEBUG("Node was skipped due to alias analysis result:\n", *node);
-      // Do NOT have enough information to do CSE on these nodes.
-      continue;
-    }
+struct CommonSubexpressionEliminator {
+  CommonSubexpressionEliminator(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
 
-    if (!node->blocks().empty()) {
-      // Traverse sub-blocks.
-      for (auto block : node->blocks()) {
-        EliminateCommonSubexpression(block, aliasDb, [&](Node* n) {
-          auto existing = subexprs.find(n);
-          if (existing != subexprs.end()) {
-            return *existing;
-          }
-
-          return parent_lookup_fn(n);
-        });
+  bool run(std::function<Node*(Node*)> parent_lookup_fn) {
+    return run(graph_->block(), std::move(parent_lookup_fn));
+  }
+
+  // The function implements common subexpression elimination.
+  // Since the nodes are visited in topological order, one pass is enough.
+  bool run(Block* block, std::function<Node*(Node*)> parent_lookup_fn) {
+    std::unordered_set<Node*, HashNode, EqualNode> subexprs;
+    bool changed = false;
+    for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
+      auto node = *it;
+      if (node->hasSideEffects()) {
+        GRAPH_DEBUG("Node was skipped due to side effects:\n", *node);
+        continue;
+      }
+      if (node->isNondeterministic()) {
+        GRAPH_DEBUG("Node was skipped due to its non determinism:\n", *node);
+        continue;
+      }
+      if (getOrCreateAliasDb().hasWriters(node)) {
```

Comment:
Can we move this after the parent_lookup check ? as is we are always going to initialize the AliasDb, it would be nice to only intiialize it if we have a CSE opportunity

Reply:
this is a bit tricky. I don't know if we can move `if (getOrCreateAliasDb().hasWriters(node))` past `if (!node->blocks().empty()) {`
since this may allow us CSE `d = a + b`.

```python
c = a + b
if (t) # returns value `d`
    d = a + b
else
    d = a + b

d.add_(3)
```

We could duplicate `hasWriters` check into `empty()` and `if (parent_lookup)` and refactor `subexprs.insert` to make sure we won't insert an expression that has writers, but it seems to be a larger refactor. I'd suggest we do it separately.

Reply:
As is, there's no point in the lazy initialization, since we're going to initialize every time. If we do it separately we should just not lazily initialize in this PR because it will never work.

I dont think we actually need to prevent CSE for blocks if the node has output writers. In your example here, even without checking if `if` has output writers, we won't try to CSE `d = a + b` because alias analysis will say that `a + b` has output writers. More generally if an if node has output writers, than the block values will also have output writers, which should prevent CSE. I dont thhink the if check does anything